### PR TITLE
PUBDEV-8880: Train XGBoost CV models concurrently on the same GPU

### PR DESCRIPTION
**Pull request by michalkurka**
Previously we tried to dedicate a GPU exclusively to one model (not just one CV model, but any model in general).

This meant that if you start one model (e.g. from Python) asynchronously and then start a second model - the first model will start training and only after it is done the second one will start executing. This is a behavior that I would not expect as a user.

There is no harm in running multiple models on GPU as long there is enough memory. While it may not be the best strategy in general, in H2O's case we have quite a bit of overhead in CPU-only scoring. Having 2 CV models running concurrently - parallelizes the overhead and saves some time 2-10% in testing - the specific speed-up depends on the frequency of scoring - for heavy scoring (e.g. every 5 trees, like AutoML) - the speed-up tends to be on the higher end, for infrequent scoring the time savings are minimal but there is never a slowdown.

I've also experimented with the (unreleased) version of XGBoost 1.6.2 integration where the speed-up is even higher.

Comparison based on Higgs 2M and Higgs 100k in different scenarios (with current XGboost):

    Test Case                        Sequential Avg Time (s)  Parallel Avg Time (s)  Relative Difference
    Small dataset, no scoring        4.8                      3.6                    1.34
    Small dataset, frequent scoring  9.9                      8.0                    1.23
    Large dataset, no scoring        26.7                     25.2                   1.06
    Large dataset, frequent scoring  73.1                     67.4                   1.08

### CODE REVIEW

File: h2o-core/src/main/java/hex/ModelBuilder.java [hunk 1]
```diff
@@ -533,11 +533,15 @@ protected int nModelsInParallel(int folds) {
   protected int nModelsInParallel(int folds, int defaultParallelization) {
     if (!_parms._parallelize_cross_validation) return 1; //user demands serial building (or we need to honor the time constraints for all CV models equally)
     int parallelization = defaultParallelization;
-    if (_train.byteSize() < 1e6) 
+    if (_train.byteSize() < smalldataSize()) 
       parallelization = folds; //for small data, parallelize over CV models
     return Math.min(parallelization, H2O.ARGS.nthreads);
   }
 
+  protected long smalldataSize() {
```
Comment (kanech) [inline h2o-core/src/main/java/hex/ModelBuilder.java:541]: Case on world boundaries, i.e. smallDataSize() ?

File: h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostGPUCVModelBuilder.java [hunk 2]
```diff
@@ -12,35 +12,77 @@ public class XGBoostGPUCVModelBuilder extends CVModelBuilder {
...
+        static int[] initUtilization(List<Integer> gpus) {
+            final int maxGpuId = gpus.stream().max(Integer::compareTo)
+                    .orElseThrow(() -> new IllegalStateException("There are no GPUs available for XGBoost (" + gpus + ")."));
+            final int[] utilization = new int[maxGpuId + 1];
+            Arrays.fill(utilization, -1);
+            gpus.forEach(id -> utilization[id] = 0);
+            return utilization;
+        }
+
+        void returnGPU(int id) {
```
Comment (kanech) [inline h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostGPUCVModelBuilder.java:68]: Not sure whether _return_(GPU) is the best antonym to _take_(LeastUtilizedGPU). IMHO, it does not remind you the internally decreased counter on the first sight. I would find _release_(GPU) more convenient naming in the context of resource usage.

Comment (michalkurka) [reply h2o-extensions/xgboost/src/main/java/hex/tree/xgboost/XGBoostGPUCVModelBuilder.java]: I think you are right, fixed. Thanks for the suggestion.

File: h2o-core/src/main/java/hex/ModelBuilder.java (hunk 1 above)
Comment (michalkurka) [reply h2o-core/src/main/java/hex/ModelBuilder.java]: Yeah, typo - "smalldata" is a one word for me because of our dataset storage ;)

**Pull request merged**